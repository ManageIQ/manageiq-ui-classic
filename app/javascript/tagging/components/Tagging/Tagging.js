/* eslint-disable react/destructuring-assignment */
import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Row, Column } from 'carbon-components-react';
import TagModifier from '../InnerComponents/TagModifier';
import TagView from '../InnerComponents/TagView';
import CategoryModifier from '../InnerComponents/CategoryModifier';
import ValueModifier from '../InnerComponents/ValueModifier';
import TaggingPropTypes from '../TaggingPropTypes';

class Tagging extends React.Component {
  // eslint-disable-next-line react/sort-comp
  onTagValueChange = (selectedTagValue) => {
    const {
      selectedTagCategory, options,
    } = this.props;
    const action = {
      tagCategory: selectedTagCategory,
      tagValue: selectedTagValue,
    };

    if (options && options.onlySingleTag) {
      this.props.onSingleTagValueChange(action, options);
    } else {
      this.props.onTagValueChange(action, options);
    }
  };

  onTagCategoryChange = (selectedTagCategory) =>
    this.props.onTagCategoryChange(selectedTagCategory);

  onTagDeleteClick = (tagCategory, tagValue) => {
    this.props.onTagDeleteClick({ tagCategory, tagValue }, this.props.options);
  };

  getCategoryValues = () =>
    (this.findSelectedTag(this.props.selectedTagCategory)
      && this.findSelectedTag(this.props.selectedTagCategory).values)
    || [];

  getSelectedCategoryValues = () =>
    this.props.assignedTags.find((tag) => tag.id === this.props.selectedTagCategory.id)
    || { values: [] };

  findSelectedTag = (selectedTagCategory = { id: undefined }) =>
    this.props.tags.find((tag) => tag.id === selectedTagCategory.id);

  isMulti = (selectedTagCategory) => {
    const selectedCategory = this.findSelectedTag(selectedTagCategory);
    if (selectedCategory && selectedCategory.singleValue) {
      return !selectedCategory.singleValue;
    }
    return true;
  }

  // eslint-disable-next-line react/destructuring-assignment
  tagCategories = this.props.tags.map((tag) => ({
    description: tag.description,
    id: tag.id,
    singleValue: tag.singleValue,
  })) || [];

  render() {
    const {
      options, selectedTagCategory, onTagCategoryChange, assignedTags,
    } = this.props;
    const isDisabled = options && options.isDisabled;
    return (
      <Grid>
        <Row>
          <Column xs={12} md={8} lg={6}>
            <TagModifier hideHeader={options && options.hideHeaders}>
              <CategoryModifier
                selectedTagCategory={selectedTagCategory}
                onTagCategoryChange={onTagCategoryChange}
                tagCategories={this.tagCategories}
                isDisabled={isDisabled}
              />
              <ValueModifier
                onTagValueChange={this.onTagValueChange}
                selectedTagValues={this.getSelectedCategoryValues().values}
                multiValue={this.isMulti(selectedTagCategory)}
                values={this.getCategoryValues()}
                isDisabled={isDisabled}
              />
            </TagModifier>
          </Column>
          <Column xs={12} md={4} lg={6}>
            <TagView
              hideHeader={options && options.hideHeaders}
              assignedTags={assignedTags}
              onTagDeleteClick={isDisabled ? () => {} : this.onTagDeleteClick}
              showCloseButton={!isDisabled}
            />
          </Column>
        </Row>
      </Grid>
    );
  }
}

Tagging.propTypes = {
  selectedTagCategory: TaggingPropTypes.category,
  tags: TaggingPropTypes.tags,
  assignedTags: TaggingPropTypes.tags,
  onTagDeleteClick: PropTypes.func.isRequired,
  onTagCategoryChange: PropTypes.func.isRequired,
  onTagValueChange: PropTypes.func.isRequired,
  onSingleTagValueChange: PropTypes.func.isRequired,
  options: PropTypes.shape({
    onlySingleTag: PropTypes.bool,
    hideHeaders: PropTypes.bool,
    isDisabled: PropTypes.bool,
  }),
};

Tagging.defaultProps = {
  selectedTagCategory: {},
  tags: [],
  assignedTags: [],
  options: {
    onlySingleTag: false,
    hideHeaders: false,
    isDisabled: false,
  },
};

export default Tagging;
