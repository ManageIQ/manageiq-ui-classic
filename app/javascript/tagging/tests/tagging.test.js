import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tagging from '../components/Tagging/Tagging';

const tags = [
  {
    description: 'Name',
    id: 1,
    values: [{ description: 'Pepa', id: 11 }, { description: 'Franta', id: 12 }],
  },
  {
    description: 'Number',
    id: 2,
    values: [{ description: '1', id: 21 }, { description: '2', id: 22 }],
  },
  {
    description: 'Animal',
    id: 3,
    values: [
      { description: 'Duck', id: 31 },
      { description: 'Cat', id: 32 },
      { description: 'Dog', id: 33 },
    ],
  },
  {
    description: 'Food',
    id: 4,
    singleValue: false,
    values: [
      { description: 'Steak', id: 41 },
      { description: 'Duck', id: 42 },
      { description: 'Salad', id: 43 },
    ],
  },
  {
    description: 'Something',
    id: 5,
    singleValue: true,
    values: [
      { description: 'Knedlik', id: 51 },
      {
        description:
          'Daenerys Stormborn of the House Targaryen, First of Her Name,...and Mother of Dragons',
        id: 52,
      },
    ],
  },
];

const selectedTagCategory = { description: 'animal', id: 1 };
const selectedTagValue = { description: 'duck', id: 1 };
const selectedTagCategory1 = { description: 'Food', id: 4 };
const selectedTagCategory2 = { description: 'Something', id: 5 };
const assignedTags = [
  {
    description: 'Name',
    id: 1,
    values: [{ description: 'Pepa', id: 11 }],
  },
];

const onChange = jest.fn();
const onDelete = jest.fn();

describe('Tagging component without redux mapping', () => {
  it('match snapshot', () => {
    const component = shallow(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onChange}
      onSingleTagValueChange={onChange}
      onTagMultiValueChange={onChange}
      onTagCategoryChange={onChange}
      onTagDeleteClick={onDelete}
      selectedTagCategory={selectedTagCategory}
      selectedTagValue={selectedTagValue}
    />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('should call methods', () => {
    const onTagCategoryChange = jest.fn();
    const onTagValueChange = jest.fn();
    const onSingleTagValueChange = jest.fn();
    const onTagDeleteClick = jest.fn();
    const onTagMultiValueChange = jest.fn();
    const wrapper = shallow(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onTagValueChange}
      onSingleTagValueChange={onSingleTagValueChange}
      onTagMultiValueChange={onTagMultiValueChange}
      onTagCategoryChange={onTagCategoryChange}
      onTagDeleteClick={onTagDeleteClick}
      selectedTagCategory={selectedTagCategory}
      selectedTagValue={selectedTagValue}
    />);
    wrapper.instance().onTagCategoryChange('xaxa');
    expect(onTagCategoryChange.mock.calls).toHaveLength(1);
    wrapper.instance().onTagValueChange('wawa');
    expect(onTagCategoryChange.mock.calls).toHaveLength(1);
    wrapper.instance().onTagDeleteClick('wowo');
    expect(onTagCategoryChange.mock.calls).toHaveLength(1);
  });

  it('should call methods - singleValue is false', () => {
    const onTagCategoryChange = jest.fn();
    const onTagValueChange = jest.fn();
    const onSingleTagValueChange = jest.fn();
    const onTagDeleteClick = jest.fn();
    const onTagMultiValueChange = jest.fn();
    const wrapper = shallow(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onTagValueChange}
      onSingleTagValueChange={onSingleTagValueChange}
      onTagMultiValueChange={onTagMultiValueChange}
      onTagCategoryChange={onTagCategoryChange}
      onTagDeleteClick={onTagDeleteClick}
      selectedTagCategory={selectedTagCategory1}
      selectedTagValue={selectedTagValue}
    />);
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(wrapper.find('ValueModifier').props().multiValue).toBe(true);
  });

  it('should call methods - singleValue is true', () => {
    const onTagCategoryChange = jest.fn();
    const onTagValueChange = jest.fn();
    const onSingleTagValueChange = jest.fn();
    const onTagDeleteClick = jest.fn();
    const onTagMultiValueChange = jest.fn();
    const wrapper = shallow(<Tagging
      tags={tags}
      assignedTags={assignedTags}
      onTagValueChange={onTagValueChange}
      onSingleTagValueChange={onSingleTagValueChange}
      onTagMultiValueChange={onTagMultiValueChange}
      onTagCategoryChange={onTagCategoryChange}
      onTagDeleteClick={onTagDeleteClick}
      selectedTagCategory={selectedTagCategory2}
      selectedTagValue={selectedTagValue}
    />);
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(wrapper.find('ValueModifier').props().multiValue).toBe(false);
  });
});
