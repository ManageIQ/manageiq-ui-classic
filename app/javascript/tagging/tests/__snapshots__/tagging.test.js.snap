// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tagging component without redux mapping match snapshot 1`] = `
<Grid>
  <Row>
    <Column
      lg={6}
      md={8}
      xs={12}
    >
      <TagModifier
        header="Add/Modify tag"
        hideHeader={false}
      >
        <CategoryModifier
          categoryLabel="Category"
          isDisabled={false}
          onTagCategoryChange={[MockFunction]}
          selectedTagCategory={
            Object {
              "description": "animal",
              "id": 1,
            }
          }
          tagCategories={
            Array [
              Object {
                "description": "Name",
                "id": 1,
                "singleValue": undefined,
              },
              Object {
                "description": "Number",
                "id": 2,
                "singleValue": undefined,
              },
              Object {
                "description": "Animal",
                "id": 3,
                "singleValue": undefined,
              },
              Object {
                "description": "Food",
                "id": 4,
                "singleValue": false,
              },
              Object {
                "description": "Something",
                "id": 5,
                "singleValue": true,
              },
            ]
          }
        />
        <ValueModifier
          isDisabled={false}
          multiValue={true}
          onTagValueChange={[Function]}
          selectedTagValues={
            Array [
              Object {
                "description": "Pepa",
                "id": 11,
              },
            ]
          }
          valueLabel="Value"
          values={
            Array [
              Object {
                "description": "Pepa",
                "id": 11,
              },
              Object {
                "description": "Franta",
                "id": 12,
              },
            ]
          }
        />
      </TagModifier>
    </Column>
    <Column
      lg={6}
      md={4}
      xs={12}
    >
      <TagView
        assignedTags={
          Array [
            Object {
              "description": "Name",
              "id": 1,
              "values": Array [
                Object {
                  "description": "Pepa",
                  "id": 11,
                },
              ],
            },
          ]
        }
        header="Assigned tags"
        hideHeader={false}
        onTagDeleteClick={[Function]}
        showCloseButton={true}
      />
    </Column>
  </Row>
</Grid>
`;

exports[`Tagging component without redux mapping should call methods - singleValue is false 1`] = `
<Grid>
  <Row>
    <Column
      lg={6}
      md={8}
      xs={12}
    >
      <TagModifier
        header="Add/Modify tag"
        hideHeader={false}
      >
        <CategoryModifier
          categoryLabel="Category"
          isDisabled={false}
          onTagCategoryChange={[MockFunction]}
          selectedTagCategory={
            Object {
              "description": "Food",
              "id": 4,
            }
          }
          tagCategories={
            Array [
              Object {
                "description": "Name",
                "id": 1,
                "singleValue": undefined,
              },
              Object {
                "description": "Number",
                "id": 2,
                "singleValue": undefined,
              },
              Object {
                "description": "Animal",
                "id": 3,
                "singleValue": undefined,
              },
              Object {
                "description": "Food",
                "id": 4,
                "singleValue": false,
              },
              Object {
                "description": "Something",
                "id": 5,
                "singleValue": true,
              },
            ]
          }
        />
        <ValueModifier
          isDisabled={false}
          multiValue={true}
          onTagValueChange={[Function]}
          selectedTagValues={Array []}
          valueLabel="Value"
          values={
            Array [
              Object {
                "description": "Steak",
                "id": 41,
              },
              Object {
                "description": "Duck",
                "id": 42,
              },
              Object {
                "description": "Salad",
                "id": 43,
              },
            ]
          }
        />
      </TagModifier>
    </Column>
    <Column
      lg={6}
      md={4}
      xs={12}
    >
      <TagView
        assignedTags={
          Array [
            Object {
              "description": "Name",
              "id": 1,
              "values": Array [
                Object {
                  "description": "Pepa",
                  "id": 11,
                },
              ],
            },
          ]
        }
        header="Assigned tags"
        hideHeader={false}
        onTagDeleteClick={[Function]}
        showCloseButton={true}
      />
    </Column>
  </Row>
</Grid>
`;

exports[`Tagging component without redux mapping should call methods - singleValue is true 1`] = `
<Grid>
  <Row>
    <Column
      lg={6}
      md={8}
      xs={12}
    >
      <TagModifier
        header="Add/Modify tag"
        hideHeader={false}
      >
        <CategoryModifier
          categoryLabel="Category"
          isDisabled={false}
          onTagCategoryChange={[MockFunction]}
          selectedTagCategory={
            Object {
              "description": "Something",
              "id": 5,
            }
          }
          tagCategories={
            Array [
              Object {
                "description": "Name",
                "id": 1,
                "singleValue": undefined,
              },
              Object {
                "description": "Number",
                "id": 2,
                "singleValue": undefined,
              },
              Object {
                "description": "Animal",
                "id": 3,
                "singleValue": undefined,
              },
              Object {
                "description": "Food",
                "id": 4,
                "singleValue": false,
              },
              Object {
                "description": "Something",
                "id": 5,
                "singleValue": true,
              },
            ]
          }
        />
        <ValueModifier
          isDisabled={false}
          multiValue={false}
          onTagValueChange={[Function]}
          selectedTagValues={Array []}
          valueLabel="Value"
          values={
            Array [
              Object {
                "description": "Knedlik",
                "id": 51,
              },
              Object {
                "description": "Daenerys Stormborn of the House Targaryen, First of Her Name,...and Mother of Dragons",
                "id": 52,
              },
            ]
          }
        />
      </TagModifier>
    </Column>
    <Column
      lg={6}
      md={4}
      xs={12}
    >
      <TagView
        assignedTags={
          Array [
            Object {
              "description": "Name",
              "id": 1,
              "values": Array [
                Object {
                  "description": "Pepa",
                  "id": 11,
                },
              ],
            },
          ]
        }
        header="Assigned tags"
        hideHeader={false}
        onTagDeleteClick={[Function]}
        showCloseButton={true}
      />
    </Column>
  </Row>
</Grid>
`;
